plugins/UsersManager/API.php

```diff
@@ -1143,34 +1144,60 @@ public function setUserAccess($userLogin, $access, $idSites, $passwordConfirmati
                 $ids = implode(', ', $this->roleProvider->getAllRoleIds());
                 throw new Exception(Piwik::translate('UsersManager_ExceptionMultipleRoleSet', $ids));
             }
+
+            $role = reset($roles);
         } else {
             // as only one access is set, we require it to be a role or "noaccess"...
             if ($access !== 'noaccess') {
                 $this->roleProvider->checkValidRole($access);
-                $roles[] = $access;
             }
+
+            $role = $access;
         }
 
         $this->checkUserExist($userLogin);
         $this->checkUsersHasNotSuperUserAccess($userLogin);
 
-        $this->model->deleteUserAccess($userLogin, $idSites);
 
-        if ($access === 'noaccess') {
+        // Get current roles per ID site
+        $idSitesAndAccess = [];
+
+        foreach ($idSites as $siteId) {
+            if ($foundAccess = $this->model->getAccessForUserForSite($userLogin, $siteId)) {
+                $idSitesAndAccess[$siteId] = $this->getRoleAndCapabilitiesFromAccess($foundAccess)[0][0];
+            }
+        }
+
+        // Re-verify admin access for current user is still applicable.
+        $this->getIdSitesCheckAdminAccess($idSites);
+
+        if ($role === 'noaccess') {
+            $this->model->deleteUserAccess($userLogin, $idSites);
             // if the access is noaccess then we don't save it as this is the default value
             // when no access are specified
             Piwik::postEvent('UsersManager.removeSiteAccess', [$userLogin, $idSites]);
         } else {
-            $role = array_shift($roles);
-            $this->model->addUserAccess($userLogin, $role, $idSites);
+            if (empty($idSitesAndAccess)) {
+                $this->model->addUserAccess($userLogin, $role, $idSites);
+            }
+
+            foreach ($idSitesAndAccess as $idSite => $previousAccess) {
+                $success = $this->model->updateUserAccessConditionally($userLogin, $idSite, $role, $previousAccess);
+                if ($success === false) {
+                    StaticContainer::get(LoggerInterface::class)->error('---- Concurrency problem ----');
+                    throw new Exception('Concurrency problem');
+                }
```

Comment:
That currently fails if there are sites that did not have an access set before, while others had access. In that case it would try to update the access on a site where no record is available. 
```suggestion
            foreach ($idSites as $idSite) {
                if (empty($idSitesAndAccess[$idSite])) {
                    $this->model->addUserAccess($userLogin, $role, $idSite);
                } else {
                    $success = $this->model->updateUserAccessConditionally($userLogin, $idSite, $role, $idSitesAndAccess[$idSite]);
                    if ($success === false) {
                        StaticContainer::get(LoggerInterface::class)->error('---- Concurrency problem ----');
                        throw new Exception('Concurrency problem');
                    }
                }
```